From 12a20ace5e553e1b1b3a020721dfbd65e6e605d5 Mon Sep 17 00:00:00 2001
From: Peng Fan <fanpeng@loongson.cn>
Date: Thu, 17 Jul 2025 12:09:03 +0000
Subject: [PATCH 04/12] LoongArch: Allow R_LARCH_PCALA_HI20 or
 R_LARCH_PCREL20_S2  against undefined weak symbols for static PIE

In a static PIE, undefined weak symbols should be just resolved to
runtime address 0, like those symbols with non-default visibility.  This
was silently broken in all prior Binutils releases with "-static-pie
-mdirect-extern-access":

    $ cat t.c
    int x (void) __attribute__ ((weak));

    int
    main (void)
    {
      __builtin_printf("%p\n", x);
    }
    $ gcc t.c -static-pie -mdirect-extern-access
    $ ./a.out
    0x7ffff1d64000

Since commit 4cb77761d687 ("LoongArch: Check PC-relative relocations for
shared libraries), the situation has been improved: the linker errors
out instead of silently producing a wrong output file.

But logically, using -mdirect-extern-access for a static PIE perfectly
makes sense, and we should not prevent that even if the programmer uses
weak symbols.  Linux kernel is such an example, and Linux < 6.10 now
fails to build with Binutils trunk.  (The silent breakage with prior
Binutils releases was "benign" due to some blind luck.)

While since the 6.10 release Linux has removed those potentially
undefined weak symbols (due to performance issue), we still should
support weak symbols in -mdirect-extern-access -static-pie and unbreak
building old kernels.
---
 bfd/elfnn-loongarch.c | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/bfd/elfnn-loongarch.c b/bfd/elfnn-loongarch.c
index 1a1a9c34..12108476 100644
--- a/bfd/elfnn-loongarch.c
+++ b/bfd/elfnn-loongarch.c
@@ -4062,12 +4062,13 @@ loongarch_elf_relocate_section (bfd *output_bfd, struct bfd_link_info *info,
 	case R_LARCH_PCALA_HI20:
 	  unresolved_reloc = false;
 
-	  /* If sym is hidden undefined weak, (sym + addend) should be
-	     resolved to runtime address (0 + addend).  */
+          /* If sym is undef weak and it's hidden or we are doing a static
+             link, (sym + addend) should be resolved to runtime address
+             (0 + addend).  */ 
 	  resolve_pcrel_undef_weak =
-	    (is_undefweak
-	     && h
-	     && ELF_ST_VISIBILITY (h->other) != STV_DEFAULT);
+            ((info->nointerp
+              || (h && ELF_ST_VISIBILITY (h->other) != STV_DEFAULT))
+             && is_undefweak);
 
 	  if (resolve_pcrel_undef_weak)
 	    pc = 0;
-- 
2.47.3

